File: python/editdistance.py
#from trie import Trie
import sys

class Trie:
    def __init__(self):
        self.children = {}
        self.is_word = False

    def insert_word(self, word):
        if len(word) == 0:
            self.is_word = True
            return
        letter = word[0]
        if letter not in self.children:
            self.children[letter] = Trie()
        self.children[letter].insert_word(word[1:])

    def contains(self, word):
        if len(word) == 0:
            return self.is_word
        letter = word[0]
        if letter in self.children:
            return self.children[letter].contains(word[1:])
        return False

def distance_helper(dict_node, letter, word, word_length, prev_row, dist):
    next_row = list(range(word_length+1))
    next_row[0] = prev_row[0]+1

    row_min = sys.maxsize
    for i in range(word_length):
        if word[i] == letter:
            cur_row_min = prev_row[i]
        else:
            cur_row_min = min(next_row[i]+1, prev_row[i]+1, prev_row[i+1]+1)
        next_row[i+1] = cur_row_min
        if cur_row_min < row_min:
            row_min = cur_row_min

    if next_row[-1] < dist and dict_node.is_word:
        dist = next_row[-1]

    if row_min < dist:
        for child in dict_node.children:
            dist = distance_helper(dict_node.children[child], child, word, word_length, next_row, dist)
            if dist == 1:
                return 1

    return dist

def distance(word, dict_words):
    if dict_words.contains(word):
        return 0

    word_length = len(word)
    row = list(range(word_length+1))
    dist = sys.maxsize
    for child in dict_words.children:
        temp_dist = distance_helper(dict_words.children[child], child, word, word_length, row, dist)
        if temp_dist == 1:
            return 1
        elif temp_dist < dist:
            dist = temp_dist

    return dist


def main():
    dictionary = Trie()

    #with open('twl06.txt', 'r') as f:
        #for line in f:
            #dict_words.insert_word(line.strip().lower())

    lines = sys.stdin.readlines()
    num_dict_words = int(lines[0])
    dict_words = [x.strip() for x in lines[:num_dict_words+1]]
    #map(lambda word: dictionary.insert_word(word), dict_words)
    for word in dict_words:
        dictionary.insert_word(word)
    words = [x.strip() for x in lines[num_dict_words+2:]]

    distances = [distance(word, dictionary) for word in words]
    print(sum(distances))


if __name__ == '__main__':
    sys.exit(main())

File: codedash2014/EditDistance.java
import java.io.*;
import java.util.LinkedList;
import java.util.List;

/**
 * Created with IntelliJ IDEA.
 * User: John
 * Date: 8/10/14
 * Time: 4:30 PM
 * To change this template use File | Settings | File Templates.
 */

class Trie
{
    public final static int MAX_CHILDREN = 26;
    public Trie()
    {
        isWord_ = false;
        children_ = new Trie[MAX_CHILDREN];
    }

    public void insertWord(String word)
    {
        if (word.length() == 0)
        {
            isWord_ = true;
            return;
        }
        int letter = word.charAt(0)-'a';
        if (children_[letter] == null)
        {
            children_[letter] = new Trie();
        }
        children_[letter].insertWord(word.substring(1));
    }

    public boolean contains(String word)
    {
        if (word.length() == 0)
        {
            return isWord_;
        }
        int letter = word.charAt(0)-'a';
        if (children_[letter] != null)
        {
            return children_[letter].contains(word.substring(1));
        }
        return false;
    }

    final Trie[] children_;
    boolean isWord_;
}

public class EditDistance {

    private static int distance_helper(Trie node, char letter, String word, int wordLength, int[] prevRow, int minDist)
    {
        int[] nextRow = new int[wordLength+1];
        nextRow[0] = prevRow[0] + 1;
        int rowMin = Integer.MAX_VALUE;
        for (int i = 0; i < wordLength; ++i)
        {
            int currentCost;
            if (letter == word.charAt(i))
            {
                currentCost = prevRow[i];
            }
            else
            {
                currentCost = Math.min(Math.min(prevRow[i]+1, prevRow[i+1]+1), nextRow[i]+1);
            }
            nextRow[i+1] = currentCost;
            if (currentCost < rowMin)
            {
                rowMin = currentCost;
            }
        }

        if (nextRow[wordLength] < minDist && node.isWord_)
        {
            minDist = nextRow[wordLength];
        }

        if (rowMin < minDist)
        {
            for (int i = 0; i < Trie.MAX_CHILDREN; ++i)
            {
                if (node.children_[i] != null)
                {
                    int dist = distance_helper(node.children_[i], (char)(i+'a'), word, wordLength, nextRow, minDist);
                    if (dist == 1)
                    {
                        return 1;
                    }
                    if (dist < minDist)
                    {
                        minDist = dist;
                    }
                }
            }
        }

        return minDist;
    }

    private static int distance(String word, Trie dict)
    {
        if (dict.contains(word))
        {
            return 0;
        }

        int minDist = Integer.MAX_VALUE;
        int wordLength = word.length();
        int[] row = new int[wordLength+1];
        for (int i = 0; i < wordLength+1; i++)
        {
            row[i] = i;
        }
        for (int i = 0; i < Trie.MAX_CHILDREN; i++)
        {
            if (dict.children_[i] != null)
            {
                int dist = distance_helper(dict.children_[i], (char)(i+'a'), word, wordLength, row, minDist);
                if (dist == 1)
                {
                    return 1;
                }
                else if (dist < minDist)
                {
                    minDist = dist;
                }
            }
        }
        return minDist;
    }

    public static void main(String[] args)
    {
        Trie dict = new Trie();
        List<String> words = new LinkedList<String>();
        try
        {
            //BufferedReader in = new BufferedReader(new FileReader("problem_2008.in"));
            BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
            int num_dict_words = Integer.parseInt(in.readLine());
            for (int i = 0; i < num_dict_words; ++i)
            {
                dict.insertWord(in.readLine());
            }
            int num_words = Integer.parseInt(in.readLine());
            for (int i = 0; i < num_words; ++i)
            {
                words.add(in.readLine());
            }
        }
        catch (FileNotFoundException e)
        {
            e.printStackTrace();
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }

        long startTime = System.currentTimeMillis();
        //int totalDistance = 0;
        StringBuilder output = new StringBuilder();
        for (String word : words)
        {
            int dist = distance(word, dict);
            //totalDistance += dist;
            //System.out.format("%s: %d\n", word, dist);
            //System.out.format("%d\n", dist);
            output.append(dist);
            output.append("\n");
        }
        long endTime = System.currentTimeMillis();

        //System.out.println(totalDistance);
        System.out.print(output.toString());
        System.err.format("took %d ms\n", (endTime-startTime));
    }
}
